### PR TITLE
Soft-delete deployments

### PR DESCRIPTION
**Pull request by jsun-m**

### CODE REVIEW

File: pkg/api/v1/deployment.go [hunk 1]
```diff
@@ -84,6 +86,11 @@ func (g *DeploymentGroup) RetrieveDeployment(ctx echo.Context) error {
 
 	deploymentId := ctx.Param("deploymentId")
 	if deployment, err := g.backendRepo.GetDeploymentByExternalId(ctx.Request().Context(), workspace.Id, deploymentId); err != nil {
+		if err == sql.ErrNoRows {
```
Comment (nickpetrovic) [inline pkg/api/v1/deployment.go:89]: This doesn't check for bad input (id not being a uuid) and will surface as a 500 when it should probably be a 400.

Check this out.
https://github.com/beam-cloud/beta9/blob/main/pkg/repository/backend_postgres.go#L419-L427

File: pkg/api/v1/deployment.go [hunk 2]
```diff
@@ -132,6 +144,44 @@ func (g *DeploymentGroup) StopAllActiveDeployments(ctx echo.Context) error {
 	return g.stopDeployments(deployments, ctx)
 }
 
+func (g *DeploymentGroup) DeleteDeployment(ctx echo.Context) error {
+	cc, ok := ctx.(*auth.HttpAuthContext)
+	if !ok {
+		return echo.NewHTTPError(http.StatusInternalServerError, "Failed to get auth info")
```
Comment (nickpetrovic) [inline pkg/api/v1/deployment.go:150]: Is this needed? I thought we took care of this [here](https://github.com/beam-cloud/beta9/blob/main/pkg/auth/http.go#L83-L98).

File: pkg/api/v1/deployment.go [hunk 3]
```diff
@@ -132,6 +144,44 @@ func (g *DeploymentGroup) StopAllActiveDeployments(ctx echo.Context) error {
...
+	}
+
+	deploymentId := ctx.Param("deploymentId")
+
+	// Get deployment
+	deploymentWithRelated, err := g.backendRepo.GetDeploymentByExternalId(ctx.Request().Context(), cc.AuthInfo.Workspace.Id, deploymentId)
+	if err != nil {
+		if err == sql.ErrNoRows {
```
Comment (nickpetrovic) [inline pkg/api/v1/deployment.go:158]: Feels like we're leaking sql into the service. Can we return `nil, nil` when we don't find a deployment and there isn't an error? See my comment about the bad sql input.

File: pkg/api/v1/deployment.go [hunk 4]
```diff
@@ -132,6 +144,44 @@ func (g *DeploymentGroup) StopAllActiveDeployments(ctx echo.Context) error {
...
+	if err := g.stopDeployments([]types.DeploymentWithRelated{*deploymentWithRelated}, ctx); err != nil {
+		return ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+			"error": "failed to stop deployment",
+		})
+	}
+
+	// Delete deployment
+	if err := g.backendRepo.DeleteDeployment(ctx.Request().Context(), deploymentWithRelated.Deployment); err != nil {
+		return ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+			"error": "failed to delete deployment",
```
Comment (nickpetrovic) [inline pkg/api/v1/deployment.go:178]: Most of these errors start lower cased, but one doesn't. Can we choose a style and keep it consistent?

File: pkg/api/v1/deployment.go (hunk 1 above)
Comment (jsun-m) [reply pkg/api/v1/deployment.go:89]: nice, will add this

File: pkg/api/v1/deployment.go (hunk 2 above)
Comment (jsun-m) [reply pkg/api/v1/deployment.go:150]: yea you are right

**Review by nickpetrovic — APPROVED** [review]

**Pull request merged**